src/coffea/nanoevents/methods/fcc.py

```diff
@@ -0,0 +1,452 @@
+            },
+            with_name="MomentumCandidate",
+            behavior=self.behavior,
+        )
+
+    @property
+    def absolute_mass(self):
+        return numpy.sqrt(numpy.abs(self.mass2))
+
+
+behavior.update(
+    awkward._util.copy_behaviors(vector.LorentzVector, MomentumCandidate, behavior)
+)
+
+MomentumCandidateArray.ProjectionClass2D = vector.TwoVectorArray  # noqa: F821
+MomentumCandidateArray.ProjectionClass3D = vector.ThreeVectorArray  # noqa: F821
+MomentumCandidateArray.ProjectionClass4D = vector.LorentzVectorArray  # noqa: F821
+MomentumCandidateArray.MomentumClass = MomentumCandidateArray  # noqa: F821
+
+
+@awkward.mixin_class(behavior)
+class MCParticle(MomentumCandidate, base.NanoCollection):
+    """Generated Monte Carlo particles"""
+
+    # Daughters
+    @dask_property
+    def get_daughters_index(self):
+        """
+        Obtain the indexes of the daughters of each and every MCParticle
+        - The output is a doubly nested awkward array
+        - Needs the presence of Particleidx1 collection
+        - The Particleidx1.index contains info about the daughters
+        """
+        ranges = index_range(self.daughters.begin, self.daughters.end)
+        return awkward.values_astype(
+            map_index_to_array(self._events().Particleidx1.index, ranges, axis=2),
+            "int64",
+        )
+
+    @get_daughters_index.dask
+    def get_daughters_index(self, dask_array):
+        """
+        Obtain the indexes of the daughters of each and every MCParticle
+        - The output is a doubly nested awkward array
+        - Needs the presence of Particleidx1 collection
+        - The Particleidx1.index contains info about the daughters
+
+        Note: Seems like all the functions need to mapped manually
+        """
+        ranges = dask_awkward.map_partitions(
+            index_range, dask_array.daughters.begin, dask_array.daughters.end
+        )
+        daughters = dask_awkward.map_partitions(
+            map_index_to_array, dask_array._events().Particleidx1.index, ranges, axis=2
+        )
+        return awkward.values_astype(daughters, "int32")
+
+    @dask_property
+    def get_daughters(self):
+        """
+        Obtain the actual MCParticle daughters to each and every MCParticle
+        - The output is a doubly nested awkward array
+        - Needs the presence of Particleidx1 collection
+        - The Particleidx1.index contains info about the daughters
+        """
+        return map_index_to_array(self, self.get_daughters_index, axis=2)
+
+    @get_daughters.dask
+    def get_daughters(self, dask_array):
+        """
+        Obtain the actual MCParticle daughters to each and every MCParticle
+        - The output is a doubly nested awkward array
+        - Needs the presence of Particleidx1 collection
+        - The Particleidx1.index contains info about the daughters
+        """
+        return map_index_to_array(dask_array, dask_array.get_daughters_index, axis=2)
```

Comment:
For this function and others below it you'll likely have the same problem we did with the NanoAOD schema where this will not work after cuts are placed on the object, since the indices are local to each event.

The solution we came up with was to reference back to the original array and then take only the objects that passed the original cuts. You can see an example for that flavor of GenParticles here: https://github.com/CoffeaTeam/coffea/blob/master/src/coffea/nanoevents/methods/nanoaod.py#L119

The main thing you have to do is generate the global indices upon reading in the file so that you can use those arrays instead of the local indices.

Reply:
Hi @lgray. Sorry for the super late reply. Indeed, this won't work if cuts are already applied. But I have trouble understanding the solution you suggested. Furthermore, I don't know what `_apply_global_index` actually does. Please explain in simpler terms how I can fix this . Thanks